eventstream/memorystream/stream.go

```diff
@@ -48,48 +89,200 @@ func (s *Stream) Open(
 	o uint64,
 	f message.TypeCollection,
 ) (eventstream.Cursor, error) {
+	if ctx.Err() != nil {
+		return nil, ctx.Err()
+	}
+
 	if f.Len() == 0 {
 		panic("at least one event type must be specified")
 	}
 
-	if ctx.Err() != nil {
-		return nil, ctx.Err()
+	// Load the tail of the linked-list first to see if the requested offset is
+	// the next expected offset. This should be the common case when used as
+	// event store cache because the memory stream is only queried when no more
+	// events can be loaded from the persistence store.
+	n := s.loadTail()
+
+	if n == nil {
+		// If the tail is nil then the list is uninitialized. We initialize it
+		// with a node that will contain the as-yet-unknown event at
+		// c.FirstOffset.
+		s.m.Lock()
+		n = s.init()
+		s.m.Unlock()
+	} else if o < n.offset {
+		// We already have an initialized list, and the requested offset is for
+		// an historical event. Starting at the head, the cursor will scan
```

Comment:
```suggestion
		// a historical event. Starting at the head, the cursor will scan
```

Reply:
Well, look at you my "I talk properly and actually pronounce the H"! :D